// victory-native/dist/utils/transform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setTranslate = exports.setScale = exports.getTransformComponents = exports.identity4 = void 0;
exports.identity4 = [
    1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
];
var MatrixValues;
(function (MatrixValues) {
    MatrixValues[MatrixValues["ScaleX"] = 0] = "ScaleX";
    MatrixValues[MatrixValues["ScaleY"] = 5] = "ScaleY";
    MatrixValues[MatrixValues["TranslateX"] = 3] = "TranslateX";
    MatrixValues[MatrixValues["TranslateY"] = 7] = "TranslateY";
})(MatrixValues || (MatrixValues = {}));
const getTransformComponents = (m) => {
    "worklet";
    return {
        scaleX: m[MatrixValues.ScaleX],
        scaleY: m[MatrixValues.ScaleY],
        translateX: m[MatrixValues.TranslateX],
        translateY: m[MatrixValues.TranslateY],
    };
};
exports.getTransformComponents = getTransformComponents;
const setScale = (matrix, kx, ky) => {
    "worklet";
    const m = matrix.slice(0);
    m[MatrixValues.ScaleX] = kx;
    m[MatrixValues.ScaleY] = ky !== null && ky !== void 0 ? ky : kx;
    return m;
};
exports.setScale = setScale;
const setTranslate = (matrix, tx, ty) => {
    "worklet";
    const m = matrix.slice(0);
    m[MatrixValues.TranslateX] = tx;
    m[MatrixValues.TranslateY] = ty;
    return m;
};
exports.setTranslate = setTranslate;
