// @shopify/react-native-skia/lib/commonjs/views/SkiaJSDomView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SkiaJSDomView = void 0;
var _react = _interopRequireDefault(require("react"));
var _Platform = require("../Platform");
var _SkiaPictureViewNativeComponent = _interopRequireDefault(require("../specs/SkiaPictureViewNativeComponent"));
var _types = require("../dom/types");
var _api = require("./api");
var _SkiaViewNativeId = require("./SkiaViewNativeId");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
const NativeSkiaPictureView = _Platform.Platform.OS !== "web" ? _SkiaPictureViewNativeComponent.default :
// eslint-disable-next-line @typescript-eslint/no-explicit-any
null;
class SkiaJSDomView extends _react.default.Component {
  constructor(props) {
    super(props);
    _defineProperty(this, "_nativeId", void 0);
    _defineProperty(this, "requestId", 0);
    this._nativeId = _SkiaViewNativeId.SkiaViewNativeId.current++;
    const {
      root,
      onSize
    } = props;
    if (root) {
      assertSkiaViewApi();
      _api.SkiaViewApi.setJsiProperty(this._nativeId, "root", root);
    }
    if (onSize) {
      assertSkiaViewApi();
      _api.SkiaViewApi.setJsiProperty(this._nativeId, "onSize", onSize);
    }
    this.tick();
  }
  tick() {
    this.redraw();
    if (this.props.mode === "continuous") {
      this.requestId = requestAnimationFrame(this.tick.bind(this));
    }
  }
  get nativeId() {
    return this._nativeId;
  }
  componentDidUpdate(prevProps) {
    const {
      root,
      onSize
    } = this.props;
    if (root !== prevProps.root && root !== undefined) {
      assertSkiaViewApi();
      this.draw();
    }
    if (onSize !== prevProps.onSize) {
      assertSkiaViewApi();
      _api.SkiaViewApi.setJsiProperty(this._nativeId, "onSize", onSize);
    }
    this.tick();
  }

  /**
   * Creates a snapshot from the canvas in the surface
   * @param rect Rect to use as bounds. Optional.
   * @returns An Image object.
   */
  makeImageSnapshot(rect) {
    assertSkiaViewApi();
    return _api.SkiaViewApi.makeImageSnapshot(this._nativeId, rect);
  }

  /**
   * Sends a redraw request to the native SkiaView.
   */
  redraw() {
    assertSkiaViewApi();
    this.draw();
  }
  draw() {
    const {
      root,
      Skia
    } = this.props;
    if (root !== undefined) {
      assertSkiaViewApi();
      const rec = Skia.PictureRecorder();
      const canvas = rec.beginRecording();
      const ctx = new _types.JsiDrawingContext(Skia, canvas);
      root.render(ctx);
      const picture = rec.finishRecordingAsPicture();
      _api.SkiaViewApi.setJsiProperty(this._nativeId, "picture", picture);
    }
  }

  /**
   * Clear up the dom node when unmounting to release resources.
   */
  componentWillUnmount() {
    assertSkiaViewApi();
    _api.SkiaViewApi.setJsiProperty(this._nativeId, "picture", null);
    if (this.requestId) {
      cancelAnimationFrame(this.requestId);
    }
  }
  render() {
    const {
      debug = false,
      ...viewProps
    } = this.props;
    return /*#__PURE__*/_react.default.createElement(NativeSkiaPictureView, _extends({
      collapsable: false,
      nativeID: `${this._nativeId}`,
      debug: debug
    }, viewProps));
  }
}
exports.SkiaJSDomView = SkiaJSDomView;
const assertSkiaViewApi = () => {
  if (_api.SkiaViewApi === null || _api.SkiaViewApi.setJsiProperty === null || _api.SkiaViewApi.requestRedraw === null || _api.SkiaViewApi.makeImageSnapshot === null) {
    throw Error("Skia View Api was not found.");
  }
};
